chaos-days/blog/2023-12-20-Broker-scaling-performance/index.md

```diff
@@ -0,0 +1,183 @@
+
+### Expected
+
+We expect that scaling up to 6 brokers will still complete successfully, even under high load.
+The time it takes until scaling is complete might be slighly higher.
+The impact on processing performance, both in terms of throughput and latency, may be slightly larger than in the previous experiment.
+
+### Actual
+
+The process instance duration did not increase, and even decreased slightly.
+![](./high_load_scaling_latency.png)
+
+Similarly, the throughput in terms of created and completed process instances and jobs remained relatively stable.
+![](./high_load_scaling_throughput.png)
+
+The number of failed requests increased slightly, but still well within an acceptable range.
+![](./high_load_scaling_failed_requests.png)
+
+The scaling operation took 5 minutes, a good portion of which is waiting for the new brokers to get scheduled and start up.
+
+Overall, this matches our expectation and shows that scaling can complete fast and with low impact on processing performance, even under high load.
+
+## Scaling up to improve performance
+
+The most obvious goal of scaling brokers is to unlock additional processing performance.
+While vertical scaling is also a great option, this can hit limits imposed by your infrastructure provider.
+For example, some machine types may offer great CPU performance but are severely limited in IOPS.
+Additionally, vertical scaling is often more expensive than horizontal scaling.
+It also comes with increased risk when a single machine fails because the remaining machines may already run at their limits and will then struggle to handle the additional load during failover.
+
+To show how broker scaling can improve processing performance, we reused the same cluster setup as before.
+We have 3 brokers, 6 partitions and replication factor 3.
+
+The brokers are limited to 1.35 CPUs and 4GiB RAM each. 
+They run with additional safety checks that are usually disabled in production and that slightly decrease the baseline processing performance.
+Each broker uses a small 32GiB SSD for storage, limiting them to a few thousand IOPS.
+
+We changed the processing load slightly to simulate a more realistic scenario.
+The new process model consists of 10 tasks with two timers in-between, each delaying the process instance by 1 minute.
+
+![](./ten_tasks.png)
+
+The processing load is generated by our own [benchmarking application](https://github.com/camunda/zeebe/tree/9e723b21b0e408fc2b97fd7d3f6b092af8e62dbe/benchmarks), initially starting 40 process instances per second.
+
+This results in 400 jobs created and completed per second.
+
+This stresses the 3 brokers and we see backpressure on all partitions.
+![](./perf_initial_backpressure.png)
+
+We also see a few jobs timing out, indicating that the cluster is unable to handle this load consistenly:
+![](./perf_initial_timeouts.png)
+
+We also see that many jobs are active for much longer than 1 second, even though the workers could complete them immediately.
```

Comment:
❓ Per default the workers have a completion delay, did you change this?

Reply:
Oh, I remember in the past the workers only used a completion delay if some "async" flag was set, which we did not do. Did that change, maybe with the job-push changes?

Reply:
No. This was always the case :)